Contents/Code/themerr_db_helper.py

```diff
@@ -0,0 +1,80 @@
+# -*- coding: utf-8 -*-
+
+# plex debugging
+try:
+    import plexhints  # noqa: F401
+except ImportError:
+    pass
+else:  # the code is running outside of Plex
+    from plexhints.log_kit import Log  # log kit
+    from plexhints.parse_kit import JSON  # parse kit
+
+# local imports
+from constants import canonical_db
+
+# imports from Libraries\Shared
+from typing import Union
+
+database_cache = {}
+
+
+def update_cache():
+    Log.Info('Updating ThemerrDB cache')
+
+    global database_cache
+    database_types = canonical_db.keys()
+
+    for database_type in database_types:
+        try:
+            pages = JSON.ObjectFromURL(
+                cacheTime=3600,
+                url='https://app.lizardbyte.dev/ThemerrDB/{}/pages.json'.format(database_type),
+                errors='ignore'  # don't crash the plugin
+            )
+            page_count = pages['pages']
+
+            id_index = set()
+
+            for page in range(page_count):
+                page_data = JSON.ObjectFromURL(
+                    cacheTime=3600,
+                    url='https://app.lizardbyte.dev/ThemerrDB/{}/all_page_{}.json'.format(database_type, page + 1),
+                    errors='ignore'  # don't crash the plugin
+                )
+
+                id_index.update(str(item['id']) for item in page_data)
+
+            database_cache[database_type] = id_index
+            Log.Info('{}: {} items in database'.format(database_type, len(id_index)))
+        except Exception as e:
+            Log.Error('{}: Error retrieving page index from ThemerrDB: {}'.format(database_type, e))
+
+
+def item_may_exist(database_type, database, id):
+    # type: (str, str, Union[int, str]) -> bool
+    """
+    Check if an item may exist in the ThemerrDB.
+
+    Parameters
+    ----------
+    database_type : str
+        The type of database to check for the item.
+
+    database : str
+        The database to check for the item.
+
+    id : Union[int, str]
+        The ID of the item to check for.
+
+    Returns
+    -------
+    bool
+        True if the item exists in the ThemerrDB (or if the cache is empty/missing), otherwise False.
+    """
+    if database != canonical_db[database_type]:
+        return True
+
+    if database_type in database_cache:
+        return str(id) in database_cache[database_type]
+
+    return True
```

Comment:
```suggestion
    if not database_type in database_cache:
        update_cache()

    return str(id) in database_cache[database_type]
```

Assuming updating this cache is pretty quick (there aren't that many pages)... How about forcing the cache to update? Then this function can be renamed to `item_exists`.

I think it should be pretty rare that this function is even called before the cache was updated at least once.

Reply:
Updating the cache (even with all pages still in the global HTTP cache) takes about 5 second on my machine (there are more than 100 pages for the movies category), I don't think doing it at each call is good. Except if we handle the backoff ourselves (make update_cache a noop if the cache was updated recently).

Why did you remove the first if (91-92)? It's there to prevent false negatives in case a database type has multiple databases with the same ID format. Say there is MovieDB1 and MovieDB2 for the type Movies. MovieX has ID 1234 in MovieDB2, but all_page_N.json returns IDs in MovieDB1. Then without that if, item_may_exist would return False even though the movie exists, just not in the page listing

Reply:
> Updating the cache (even with all pages still in the global HTTP cache) takes about 5 second on my machine (there are more than 100 pages for the movies category), I don't think doing it at each call is good.

How about forcing a cache update as part of the plugin `Start` function? Although it should already be run very early in the plugin initialization phase since it's part of the scheduled task, which runs once when the plugin first starts up.

Is my thinking correct that the cache will be available for the life of the plugin, once it has run at least one time?

Reply:
> Is my thinking correct that the cache will be available for the life of the plugin, once it has run at least one time?

Yes

> Although it should already be run very early in the plugin initialization phase since it's part of the scheduled task, which runs once when the plugin first starts up.

Yes, and as such I'm not sure it would be useful to also run it in the `Start` function (since it will be run by the scheduled theme update task anyway), but I can add it if you're really sure